### PR TITLE
Optimize no-op span creation

### PR DESCRIPTION
**Pull request by pyohannes**
Don't create a `NoopSpan` for each `NoopTracer::StartSpan` call or for each `DROP` sampling decision, but instead use a static singleton in both cases.

This causes a significant performance improvement for the [BM_NoopSpanCreation](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/test/trace/sampler_benchmark.cc#L173) benchmark.

### CODE REVIEW

**Review by lalitb — APPROVED** [review]

File: api/include/opentelemetry/trace/noop.h [hunk 1]
```diff
@@ -66,7 +66,12 @@ class NoopTracer final : public Tracer, public std::enable_shared_from_this<Noop
                                     const KeyValueIterable & /*attributes*/,
                                     const StartSpanOptions & /*options*/) noexcept override
   {
-    return nostd::shared_ptr<Span>{new (std::nothrow) NoopSpan{this->shared_from_this()}};
+    // Don't allocate a no-op span for every StartSpan call, but use a static
+    // singleton for this case.
+    static nostd::shared_ptr<trace_api::Span> noop_span(
+        new trace_api::NoopSpan{this->shared_from_this()});
+
+    return noop_span;
```
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/noop.h:74]: Wondering whether there could be issue if the caller releases the returned `shared_ptr`.

Comment (ThomsonTan) [reply api/include/opentelemetry/trace/noop.h:74]: It seems ref-count is not an issue here. Could we use `static auto noop_span(std::make_shared<trace_api::NoopSpan>(this->shared_from_this()))` which could save 1 heap allocation (the span and shared_ptr will be allocated in one heap allocation call).

Comment (pyohannes) [reply api/include/opentelemetry/trace/noop.h:74]: We need a `nostd::shared_ptr` here, not a `std::shared_ptr`. So I'm afraid using `make_shared` will not benefit us here.

Comment (ThomsonTan) [reply api/include/opentelemetry/trace/noop.h:74]: Ok. It makes sense if `nostd::shared_ptr` has to be used here.

**Review by reyang — APPROVED** [review]

**Pull request merged**